File: tests/Aiko.Core.Tests/SkillPluginTests.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Aiko.Core.Tests;

public class SkillPluginTests
{
    private static (string, byte[]) File(string path, string text) => (path, Encoding.UTF8.GetBytes(text));

    [Fact]
    public void The_hash_changes_with_any_file_and_not_with_the_order_or_the_slashes()
    {
        var files = new[] { File("skills/copy/SKILL.md", "a"), File(".claude-plugin/plugin.json", "{}") };

        var hash = SkillPlugin.ContentHash(files);

        Assert.Matches("^[0-9a-f]{12}$", hash);
        Assert.Equal(hash, SkillPlugin.ContentHash(files.Reverse()));
        Assert.Equal(hash, SkillPlugin.ContentHash([File(@"skills\copy\SKILL.md", "a"), File(@".claude-plugin\plugin.json", "{}")]));
        Assert.NotEqual(hash, SkillPlugin.ContentHash([File("skills/copy/SKILL.md", "b"), File(".claude-plugin/plugin.json", "{}")]));
    }

    [Fact]
    public void The_copy_carries_the_plugin_version_with_the_hash()
    {
        var manifest = SkillPlugin.VersionedManifest("""{ "name": "aiko", "version": "1.2.0", "license": "Apache-2.0" }""", "0123456789ab");

        var root = JsonNode.Parse(manifest!)!.AsObject();
        Assert.Equal("1.2.0+0123456789ab", root["version"]!.GetValue<string>());
        Assert.Equal("Apache-2.0", root["license"]!.GetValue<string>());
        Assert.Equal("1.2.0+0123456789ab", SkillPlugin.VersionOf(manifest));

        // Copied again, the old hash is replaced, not added to.
        Assert.Equal("1.2.0+ffffffffffff", SkillPlugin.VersionOf(SkillPlugin.VersionedManifest(manifest!, "ffffffffffff")));
        Assert.Equal("0.0.0+0123456789ab", SkillPlugin.VersionOf(SkillPlugin.VersionedManifest("""{ "name": "aiko" }""", "0123456789ab")));
        Assert.Null(SkillPlugin.VersionedManifest("not json", "0123456789ab"));
    }

    [Fact]
    public void Only_the_skills_plugin_is_read_from_a_manifest()
    {
        Assert.Equal(new SkillPlugin("aiko", "Text."), SkillPlugin.FromManifest("""{ "name": "aiko", "description": "Text." }"""));
        Assert.Null(SkillPlugin.FromManifest("""{ "name": "aiko-copy" }"""));
        Assert.Null(SkillPlugin.FromManifest("""{ "name": "someone-else" }"""));
        Assert.Null(SkillPlugin.FromManifest(null));
    }

    [Fact]
    public void The_marketplace_lists_the_persona_and_then_the_skills_plugin()
    {
        var json = AikoMarketplace.Json("\"x.exe\" plugin aiko-persona", new SkillPlugin("aiko", "All skills."));

        var plugins = JsonNode.Parse(json)!["plugins"]!.AsArray();
        Assert.Equal(["aiko-persona", "aiko"], plugins.Select(p => p!["name"]!.GetValue<string>()));
        Assert.Equal("./plugins/aiko", plugins[1]!["source"]!.GetValue<string>());
        Assert.Single(JsonNode.Parse(AikoMarketplace.Json("\"x.exe\" plugin aiko-persona"))!["plugins"]!.AsArray());
    }

    // ---- the plugin in the repository ----

    private static string Repository()
    {
        var folder = new DirectoryInfo(AppContext.BaseDirectory);
        while (folder is not null && !System.IO.File.Exists(Path.Combine(folder.FullName, "Aiko.slnx")))
        {
            folder = folder.Parent;
        }

        return folder?.FullName ?? throw new InvalidOperationException("the repository root was not found");
    }

    private static string PluginFolder() => Path.Combine(Repository(), SkillPlugin.FolderName, SkillCatalog.PluginName);

    public static TheoryData<string> ShippedSkills()
    {
        var data = new TheoryData<string>();
        foreach (var folder in Directory.GetDirectories(Path.Combine(PluginFolder(), SkillPlugin.SkillsFolder)))
        {
            data.Add(Path.GetFileName(folder));
        }

        return data;
    }

    [Fact]
    public void One_plugin_ships_and_it_holds_every_skill_of_the_catalog()
    {
        Assert.Equal([SkillCatalog.PluginName], Directory.GetDirectories(Path.Combine(Repository(), SkillPlugin.FolderName)).Select(Path.GetFileName));

        var manifest = SkillPlugin.FromManifest(System.IO.File.ReadAllText(Path.Combine(PluginFolder(), ".claude-plugin", "plugin.json")));
        Assert.Equal(SkillCatalog.PluginName, manifest?.Name);
        Assert.False(string.IsNullOrWhiteSpace(manifest?.Description));

        var folders = Directory.GetDirectories(Path.Combine(PluginFolder(), SkillPlugin.SkillsFolder)).Select(Path.GetFileName).Order(StringComparer.Ordinal);
        Assert.Equal(SkillCatalog.All.Order(StringComparer.Ordinal), folders);
    }

    [Theory]
    [MemberData(nameof(ShippedSkills))]
    public void A_shipped_skill_is_named_after_its_folder(string name)
    {
        var skill = System.IO.File.ReadAllText(Path.Combine(PluginFolder(), SkillPlugin.SkillsFolder, name, "SKILL.md")).ReplaceLineEndings("\n");

        Assert.StartsWith("---", skill);
        Assert.Contains($"\nname: {name}\n", skill);
        Assert.Contains("\ndescription: ", skill);
        Assert.DoesNotContain("slayumind", skill, StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public void The_public_marketplace_lists_the_one_plugin()
    {
        using var json = JsonDocument.Parse(System.IO.File.ReadAllText(Path.Combine(Repository(), ".claude-plugin", "marketplace.json")));
        var listed = json.RootElement.GetProperty("plugins").EnumerateArray()
            .Select(p => (p.GetProperty("name").GetString(), p.GetProperty("source").GetString()))
            .ToList();

        Assert.Equal([(SkillCatalog.PluginName, $"./plugins/{SkillCatalog.PluginName}")], listed);
    }
}

File: mac/Tests/AikoKitTests/SkillPluginTests.swift
import Foundation
import Testing

@testable import AikoKit

struct SkillPluginTests {
    private func file(_ path: String, _ text: String) -> (path: String, content: Data) {
        (path, Data(text.utf8))
    }

    @Test
    func theHashChangesWithAnyFileAndNotWithTheOrderOrTheSlashes() {
        let files = [file("skills/copy/SKILL.md", "a"), file(".claude-plugin/plugin.json", "{}")]

        let hash = SkillPlugin.contentHash(files)

        #expect(hash.count == 12)
        #expect(hash.allSatisfy { $0.isHexDigit && !$0.isUppercase })
        #expect(hash == SkillPlugin.contentHash(files.reversed()))
        #expect(
            hash
                == SkillPlugin.contentHash([
                    file(#"skills\copy\SKILL.md"#, "a"), file(#".claude-plugin\plugin.json"#, "{}"),
                ]))
        #expect(
            hash
                != SkillPlugin.contentHash([
                    file("skills/copy/SKILL.md", "b"), file(".claude-plugin/plugin.json", "{}"),
                ]))
    }

    @Test
    func theCopyCarriesThePluginVersionWithTheHash() throws {
        let manifest = try #require(
            SkillPlugin.versionedManifest(#"{ "name": "aiko", "version": "1.2.0", "license": "Apache-2.0" }"#, "0123456789ab"))

        let root = try #require(JsonNode.parse(manifest)?.objectValue)
        #expect(root["version"]?.stringValue == "1.2.0+0123456789ab")
        #expect(root["license"]?.stringValue == "Apache-2.0")
        #expect(SkillPlugin.versionOf(manifest) == "1.2.0+0123456789ab")

        // Copied again, the old hash is replaced, not added to.
        #expect(SkillPlugin.versionOf(SkillPlugin.versionedManifest(manifest, "ffffffffffff")) == "1.2.0+ffffffffffff")
        #expect(
            SkillPlugin.versionOf(SkillPlugin.versionedManifest(#"{ "name": "aiko" }"#, "0123456789ab"))
                == "0.0.0+0123456789ab")
        #expect(SkillPlugin.versionedManifest("not json", "0123456789ab") == nil)
    }

    @Test
    func onlyTheSkillsPluginIsReadFromAManifest() {
        #expect(
            SkillPlugin.fromManifest(#"{ "name": "aiko", "description": "Text." }"#)
                == SkillPlugin(name: "aiko", description: "Text."))
        #expect(SkillPlugin.fromManifest(#"{ "name": "aiko-copy" }"#) == nil)
        #expect(SkillPlugin.fromManifest(#"{ "name": "someone-else" }"#) == nil)
        #expect(SkillPlugin.fromManifest(nil) == nil)
    }

    @Test
    func theMarketplaceListsThePersonaAndThenTheSkillsPlugin() {
        let json = AikoMarketplace.json(
            #""x.exe" plugin aiko-persona"#, skills: SkillPlugin(name: "aiko", description: "All skills."))

        let plugins = JsonNode.parse(json)?["plugins"]?.arrayValue ?? []
        #expect(plugins.compactMap { $0["name"]?.stringValue } == ["aiko-persona", "aiko"])
        #expect(plugins[1]["source"]?.stringValue == "./plugins/aiko")
        #expect(JsonNode.parse(AikoMarketplace.json(#""x.exe" plugin aiko-persona"#))?["plugins"]?.arrayValue?.count == 1)
    }

    // ---- the plugin in the repository ----

    static func repository() -> URL {
        var folder = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        while folder.path != "/" && !FileManager.default.fileExists(atPath: folder.appendingPathComponent("Aiko.slnx").path) {
            folder = folder.deletingLastPathComponent()
        }
        return folder
    }

    static func pluginFolder() -> URL {
        repository().appendingPathComponent(SkillPlugin.folderName).appendingPathComponent(SkillCatalog.pluginName)
    }

    static func shippedSkills() -> [String] {
        let skills = pluginFolder().appendingPathComponent(SkillPlugin.skillsFolder)
        return ((try? FileManager.default.contentsOfDirectory(atPath: skills.path)) ?? [])
            .filter { !$0.hasPrefix(".") }
            .sorted()
    }

    @Test
    func onePluginShipsAndItHoldsEverySkillOfTheCatalog() throws {
        let plugins = try FileManager.default.contentsOfDirectory(
            atPath: Self.repository().appendingPathComponent(SkillPlugin.folderName).path)
        #expect(plugins.filter { !$0.hasPrefix(".") } == [SkillCatalog.pluginName])

        let manifestPath = Self.pluginFolder().appendingPathComponent(".claude-plugin/plugin.json")
        let manifest = SkillPlugin.fromManifest(try String(contentsOf: manifestPath, encoding: .utf8))
        #expect(manifest?.name == SkillCatalog.pluginName)
        let description = (manifest?.description ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        #expect(description.isEmpty == false)

        #expect(Self.shippedSkills() == SkillCatalog.all.sorted())
    }

    @Test(arguments: SkillPluginTests.shippedSkills())
    func aShippedSkillIsNamedAfterItsFolder(name: String) throws {
        let path = Self.pluginFolder()
            .appendingPathComponent(SkillPlugin.skillsFolder)
            .appendingPathComponent(name)
            .appendingPathComponent("SKILL.md")
        let skill = try String(contentsOf: path, encoding: .utf8).replacingOccurrences(of: "\r\n", with: "\n")

        #expect(skill.hasPrefix("---"))
        #expect(skill.contains("\nname: \(name)\n"))
        #expect(skill.contains("\ndescription: "))
        #expect(!skill.lowercased().contains("slayumind"))
    }

    @Test
    func thePublicMarketplaceListsTheOnePlugin() throws {
        let path = Self.repository().appendingPathComponent(".claude-plugin/marketplace.json")
        let json = try #require(JsonNode.parse(try String(contentsOf: path, encoding: .utf8)))
        let listed = (json["plugins"]?.arrayValue ?? []).map {
            ($0["name"]?.stringValue, $0["source"]?.stringValue)
        }

        #expect(listed.count == 1)
        #expect(listed.first?.0 == SkillCatalog.pluginName)
        #expect(listed.first?.1 == "./plugins/\(SkillCatalog.pluginName)")
    }
}
